File: Tests/test_flash_vle.m
function test_flash_vle()
% test_flash_vle  Integration tests for VLE flash.
%
% System: methanol (CH4O) + water (H2O) at 322.91 K with kij = -0.07.
% The negative flash traces the phase envelope using equimolar feed; vapor
% fractions outside [0,1] are expected when the feed is in the single-phase
% region (that is the purpose of the negative saturation method).
%
% We test:
%   1. vleflashnegative produces finite, normalised compositions.
%   2. At p = 26131 Pa (equimolar feed is two-phase) the vapor composition
%      matches experiment within 0.04 mole fraction.
%   3. Mass balance: z = V*y + (1-V)*x for a two-phase flash.
%   4. vleflash (standard) gives valid results in [0,1] for a two-phase feed.
%   5. kval_estimate produces positive K-values.

[component, ~] = addComponents({'CH4O', 'H2O'});
T0      = 322.91;          % [K]
thermo1 = addThermo();
thermo1.EOS = @PREOS;
options = FlashOptions();

% --- 1. Negative flash: compositions are finite and sum to 1 ---------------
% Three representative experimental pressures (Pa) from Ochi et al. (1986).
p_exp   = [15932, 26131, 52142];
y1_exp  = [0.2741, 0.6294, 0.9736];  % vapor methanol mole fraction

for i = 1:3
    mix = Mixture(component, T0, p_exp(i));
    mix.bip.EOScons = [0 -0.07; -0.07 0];
    [vapor_y, liquid_x, vapor_frac] = vleflashnegative(mix, thermo1, options);

    pvt_assert(isfinite(vapor_frac), ...
        sprintf('vleflashnegative p=%d: vapor_frac is finite', p_exp(i)));
    pvt_assert(isfinite(sum(vapor_y)) && isfinite(sum(liquid_x)), ...
        sprintf('vleflashnegative p=%d: compositions are finite', p_exp(i)));
    pvt_assert(abs(sum(vapor_y) - 1) < 1e-4, ...
        sprintf('vleflashnegative p=%d: vapor_y sums to 1', p_exp(i)));
    pvt_assert(abs(sum(liquid_x) - 1) < 1e-4, ...
        sprintf('vleflashnegative p=%d: liquid_x sums to 1', p_exp(i)));
end

% --- 2. Vapor composition matches experiment at middle pressure --------
% At p=26131 Pa with equimolar feed the vapor fraction is ~0.71 (two-phase);
% the vapor composition should agree with experiment within 0.04.
TOL_y = 0.04;
mix_mid = Mixture(component, T0, p_exp(2));
mix_mid.bip.EOScons = [0 -0.07; -0.07 0];
[vapor_y_mid, liquid_x_mid, vf_mid] = vleflashnegative(mix_mid, thermo1, options);

pvt_assert(vf_mid >= 0 && vf_mid <= 1, ...
    'vleflashnegative p=26131: equimolar feed is two-phase (vf in [0,1])');
pvt_assert(abs(vapor_y_mid(1) - y1_exp(2)) < TOL_y, ...
    sprintf('vleflashnegative p=26131: y(methanol) within %.2f of experiment', TOL_y));

% --- 3. Mass balance check (z = V*y + (1-V)*x) --------------------------
z = mix_mid.mole_fraction;
z_check = vf_mid * vapor_y_mid(1) + (1 - vf_mid) * liquid_x_mid(1);
pvt_assert(abs(z_check - z(1)) < 1e-3, 'vleflashnegative: mass balance satisfied');

% --- 4. Standard vleflash: two-phase feed with known compositions --------
% Use a feed composition (z1=0.35) that lies between x1=0.2131 and y1=0.6294
% at p=26131 Pa, guaranteeing a two-phase result.
mix_tp = Mixture(component, T0, 26131);
mix_tp.mole_fraction = [0.35 0.65];
mix_tp.bip.EOScons = [0 -0.07; -0.07 0];
[vy_tp, lx_tp, vf_tp] = vleflash(mix_tp, thermo1, options);

pvt_assert(vf_tp >= 0 && vf_tp <= 1,     'vleflash: vapor fraction in [0,1]');
pvt_assert(abs(sum(vy_tp) - 1) < 1e-6,   'vleflash: vapor_y sums to 1');
pvt_assert(abs(sum(lx_tp) - 1) < 1e-6,   'vleflash: liquid_x sums to 1');
pvt_assert(all(vy_tp >= 0),               'vleflash: vapor_y non-negative');
pvt_assert(all(lx_tp >= 0),               'vleflash: liquid_x non-negative');
% Check mass balance
z_tp = mix_tp.mole_fraction(1);
z_tp_check = vf_tp * vy_tp(1) + (1 - vf_tp) * lx_tp(1);
pvt_assert(abs(z_tp_check - z_tp) < 1e-4, 'vleflash: mass balance satisfied');

% --- 5. kval_estimate gives positive K-values ---------------------------
K = kval_estimate(mix_tp);
pvt_assert(numel(K) == 2, 'kval_estimate: returns N K-values');
pvt_assert(all(K > 0),    'kval_estimate: all K-values positive');

File: tests/test_flash_vle.py
"""Integration tests for VLE flash."""

import numpy as np
import pytest

from pvttool import (
    FlashOptions,
    add_components,
    add_mixture,
    add_thermo,
    vle_flash,
    vle_flash_negative,
)
from pvttool.flash._kvalue import kval_estimate


# Experimental data: Ochi et al. (1986) methanol-water at 322.91 K
_P_EXP = [15932.0, 26131.0, 52142.0]
_Y1_EXP = [0.2741, 0.6294, 0.9736]  # vapor methanol mole fraction


@pytest.fixture
def mw_comps():
    return add_components(["Methanol", "Water"])


def _make_mw_mix(comps, p):
    mix = add_mixture(comps, T=322.91, p=p)
    mix.bip.eos_cons = np.array([[0.0, -0.07], [-0.07, 0.0]])
    return mix


def test_vle_flash_negative_finite(mw_comps):
    """vle_flash_negative produces finite, normalised compositions."""
    thermo = add_thermo()
    opts = FlashOptions()
    for p_val in _P_EXP:
        mix = _make_mw_mix(mw_comps, p_val)
        y, x, V = vle_flash_negative(mix, thermo, opts)
        assert np.isfinite(V)
        assert np.isfinite(y.sum())
        assert np.isfinite(x.sum())
        assert abs(y.sum() - 1.0) < 1e-4
        assert abs(x.sum() - 1.0) < 1e-4


def test_vle_flash_negative_two_phase(mw_comps):
    """At p=26131 Pa equimolar feed is two-phase; vapor fraction in [0,1]."""
    mix = _make_mw_mix(mw_comps, _P_EXP[1])
    thermo = add_thermo()
    y, x, V = vle_flash_negative(mix, thermo)
    assert 0.0 <= V <= 1.0


def test_vle_flash_negative_vapor_composition(mw_comps):
    """Vapor methanol fraction matches experiment within 0.05 at p=26131 Pa."""
    mix = _make_mw_mix(mw_comps, _P_EXP[1])
    thermo = add_thermo()
    y, x, V = vle_flash_negative(mix, thermo)
    assert abs(y[0] - _Y1_EXP[1]) < 0.05


def test_vle_flash_mass_balance(mw_comps):
    """z = V*y + (1-V)*x at p=26131 Pa."""
    mix = _make_mw_mix(mw_comps, _P_EXP[1])
    thermo = add_thermo()
    y, x, V = vle_flash_negative(mix, thermo)
    z = mix.mole_fraction
    z_check = V * y + (1.0 - V) * x
    np.testing.assert_allclose(z_check, z, atol=1e-3)


def test_vle_flash_standard_two_phase(mw_comps):
    """vle_flash with z1=0.35 gives vapor_frac in [0,1] and valid compositions."""
    mix = _make_mw_mix(mw_comps, 26131.0)
    mix.mole_fraction = np.array([0.35, 0.65])
    thermo = add_thermo()
    opts = FlashOptions()
    y, x, V = vle_flash(mix, thermo, opts)
    assert 0.0 <= V <= 1.0
    assert abs(y.sum() - 1.0) < 1e-6
    assert abs(x.sum() - 1.0) < 1e-6
    assert all(y >= 0)
    assert all(x >= 0)
    # mass balance
    z = mix.mole_fraction[0]
    z_check = V * y[0] + (1.0 - V) * x[0]
    assert abs(z_check - z) < 1e-4


def test_kval_estimate_positive(mw_comps):
    mix = _make_mw_mix(mw_comps, 26131.0)
    K = kval_estimate(mix)
    assert len(K) == 2
    assert all(K > 0)


def test_vle_flash_pure_component_trivial():
    """Pure-component flash should return trivial result (V=0 or 1)."""
    comps = add_components(["Methane"])
    mix = add_mixture(comps, T=200.0, p=1e5)
    mix.mole_fraction = np.array([1.0])
    thermo = add_thermo()
    y, x, V = vle_flash(mix, thermo)
    assert abs(y.sum() - 1.0) < 1e-6
    assert abs(x.sum() - 1.0) < 1e-6
